npm/networkPolicyController.go

```diff
@@ -0,0 +1,498 @@
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(c.runWorker, time.Second, stopCh)
+	}
+
+	klog.Infof("Started Network Policy %d worker(s)", threadiness)
+	<-stopCh
+	klog.Info("Shutting down Network Policy workers")
+
+	return nil
+}
+
+func (c *networkPolicyController) runWorker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *networkPolicyController) processNextWorkItem() bool {
+	obj, shutdown := c.workqueue.Get()
+
+	if shutdown {
+		return false
+	}
+
+	err := func(obj interface{}) error {
+		defer c.workqueue.Done(obj)
+		var key string
+		var ok bool
+		if key, ok = obj.(string); !ok {
+			// As the item in the workqueue is actually invalid, we call
+			// Forget here else we'd go into a loop of attempting to
+			// process a work item that is invalid.
+			c.workqueue.Forget(obj)
+			utilruntime.HandleError(fmt.Errorf("expected string in workqueue but got %#v", obj))
+			return nil
+		}
+		// Run the syncNetPol, passing it the namespace/name string of the
+		// network policy resource to be synced.
+		// TODO : may consider using "c.queue.AddAfter(key, *requeueAfter)" according to error type later
+		if err := c.syncNetPol(key); err != nil {
+			// Put the item back on the workqueue to handle any transient errors.
+			c.workqueue.AddRateLimited(key)
+			return fmt.Errorf("error syncing '%s': %s, requeuing", key, err.Error())
+		}
+		// Finally, if no error occurs we Forget this item so it does not
+		// get queued again until another change happens.
+		c.workqueue.Forget(obj)
+		klog.Infof("Successfully synced '%s'", key)
+		return nil
+	}(obj)
+
+	if err != nil {
+		utilruntime.HandleError(err)
+		metrics.SendErrorLogAndMetric(util.NetpolID, "syncNetPol error due to %v", err)
+		return true
+	}
+
+	return true
+}
+
+// syncNetPol compares the actual state with the desired, and attempts to converge the two.
+func (c *networkPolicyController) syncNetPol(key string) error {
+	// Convert the namespace/name string into a distinct namespace and name
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("invalid resource key: %s", key))
+		return nil
+	}
+
+	// Get the network policy resource with this namespace/name
+	netPolObj, err := c.netPolLister.NetworkPolicies(namespace).Get(name)
+
+	// (TODO): Reduce scope of lock later
+	c.npMgr.Lock()
+	defer c.npMgr.Unlock()
+
+	if err != nil {
+		if errors.IsNotFound(err) {
+			klog.Infof("Network Policy %s is not found, may be it is deleted", key)
+			// netPolObj is not found, but should need to check the RawNpMap cache with cachedNetPolKey
+			// #1 No item in RawNpMap, which means network policy with the cachedNetPolKey is not applied. Thus, do not need to clean up process.
+			// #2 item in RawNpMap exists, which means network policy with the cachedNetPolKey is applied . Thus, Need to clean up process.
+			cachedNetPolKey := util.GetNSNameWithPrefix(key)
+			_, cachedNetPolObjExists := c.npMgr.RawNpMap[cachedNetPolKey]
```

Comment:
Do we have to check if the cahedObj exists here ? Because the first check we do in cleanUpnetworkPolicy is if the cachedObj exists, i think we can safely remove this check. Any reason for leaving this check here ?

Reply:
Good catch. Removed the redundant check.